XenAdmin/Dialogs/AttachUsbDialog.cs

```diff
@@ -56,16 +56,48 @@ private void BuildList()
             treeUsbList.BeginUpdate();
             try
             {
-                Host[] hosts = _vm.Connection.Cache.Hosts;
-                foreach (Host host in hosts)
+                VM.HA_Restart_Priority SelectedPriority = _vm.HARestartPriority();
+                Pool pool = Helpers.GetPool(_vm.Connection);
+                // Check if HA was enabled on pool and Restart priority was set on VM.
+                bool haEnabled = (pool != null &&
+                    pool.ha_enabled &&
+                    VM.HaPriorityIsRestart(_vm.Connection, SelectedPriority));
+                List<XenRef<Host>> possibleHostRefs = VM.get_possible_hosts(_vm.Connection.Session, _vm.opaque_ref);
+                List<Host> possibleHosts = new List<Host>();
+                foreach (XenRef<Host> possibleHostRef in possibleHostRefs)
+                {
+                    possibleHosts.Add(_vm.Connection.Resolve(possibleHostRef));
+                }
+
+                foreach (Host host in possibleHosts)
                 {
+                    // Add a host node to tree list.
                     HostItem hostNode = new HostItem(host);
                     treeUsbList.AddNode(hostNode);
                     List<PUSB> pusbs = _vm.Connection.ResolveAll(host.PUSBs);
                     foreach (PUSB pusb in pusbs)
                     {
-                        UsbItem usbNode = new UsbItem(pusb);
-                        treeUsbList.AddChildNode(hostNode, usbNode);
+                        // Add a USB in the host to tree list.
+                        // Determin if the USB is valid to attach.
+                        if (pusb != null)
+                        {
+                            bool attached = false;
+                            USB_group usbGroup = pusb.Connection.Resolve(pusb.USB_group);
+                            if ((usbGroup != null) && 
+                                (usbGroup.VUSBs != null) && 
+                                (usbGroup.VUSBs.Count > 0))
+                                attached = true;
+                            else
+                                attached = false;
+
+                            if ((haEnabled == false) &&
+                                (pusb.passthrough_enabled == true) &&

```

Comment:
This can be written: 
`if (pusb.passthrough_enabled && !attached)` 
 (I ignored haEnabled as we agreed that this dialog won't even be shown is ha is enabled)